### PR TITLE
Resolve #1436: Add AWS Golang template, tests, and instructions on simple quick start

### LINKED ISSUE #1436: Go support
It would be great to implement Go support (maybe through a shim).

### PR DESCRIPTION
**Pull request by license2e**
Pull Request Template ------------------------
* Check out our Pull Request Guidelines here: https://github.com/serverless/serverless/blob/v1.0/CONTRIBUTING.md

##### Status:

Ready/In Development/Hold

##### Todos:

- [x] Write tests

Signed-off-by: Ilya Shindyapin <ilya@shindyapin.com>

### CODE REVIEW

File: lib/plugins/create/templates/README.md [hunk 1]
```diff
@@ -49,3 +47,15 @@ afterwards.
...
 7. `serverless remove`
+
+## Golang
+
+Follow these simple steps to create and deploy your golang service, run your function and remove the service afterwards.
+
+1. install latest version of `serverless` and `go`
+2. `mkdir my-first-service && cd my-first-service`
+3. `serverless create --template aws-golang`
+4. `GOOS=linux GOARCH=amd64 go build -o hello hello.go`
```
Comment (eahefnawy) [inline lib/plugins/create/templates/README.md]: Do users have to set those env vars? Is there any way around it?

Comment (flomotlik) [inline lib/plugins/create/templates/README.md]: Thats just standard Go Build stuff. If you have a build tool defined for example it would take care of this. But for a Go developer this is not surprising and totally standard.

**Comment by mthenw** [comment]
My opinion [haven't changed](https://github.com/serverless/serverless/pull/1680#issuecomment-235504283). Would be great to have Golang support but I'm not sure if it's worth to add it as it's basically a hack. And it will be a hack until AWS will not add native support.

I saw that you added error handling. But there is another issue. There is no way to log something from function written in Golang. You can't send it to stdout because it's used for getting result from functions. You can't use stderr neither because it's used for getting error from functions. 

Still, it's great that you created that template. Maybe we should consider adding support for third-party templates. Something like ```serverless create --template=github.com/license2e/serverless-golang-template```. So contributors can store them in separate repos.

**Comment by flomotlik** [comment]
From a product perspective I would really like to have experimental Go support in Serverless until AWS actually releases full Go support in Lambda.

This would help us in various ways:

1. We learn more about the needs of the Go community on Lambda before building a full integration
2. Make Serverless attractive for Go developers who currently have to look at Apex or others

So I'm in favour of getting this into Serverless as a preparation until we do full Go support. A few caveats though we have to make sure are in place:

1. We need to clearly mark this as experimental (e.g. experimental-aws-go as the template name)
2. The shim needs to support returning arbitrary JSON (currently it returns with a specific message) that it gets through STDOUT
3. The shim needs to support logging through STDERR (STDOUT is used for returning JSON)
4. The shim exits correctly depending on the Exit status of the Go program (which it already does I think)

Basically on a shim level it has to work the same as apex-go does.

With that in Place I would really like this template to be included in the core and moved to official support once AWS launches it. Thanks @license2e for getting this started!

File: lib/plugins/create/create.js [hunk 2]
```diff
@@ -9,6 +9,7 @@ const validTemplates = [
   'aws-python',
   'aws-java-maven',
   'aws-java-gradle',
+  'aws-golang',
```
Comment (flomotlik) [inline lib/plugins/create/create.js]: change to experimental-aws-golang

File: lib/plugins/create/templates/README.md [hunk 3]
```diff
@@ -49,3 +47,15 @@ afterwards.
 5. `serverless deploy`
 6. `serverless invoke --function hello --path event.json`
 7. `serverless remove`
+
+## Golang
+
+Follow these simple steps to create and deploy your golang service, run your function and remove the service afterwards.
```
Comment (flomotlik) [inline lib/plugins/create/templates/README.md]: Add a comment describing that this is experimental

File: lib/plugins/create/templates/aws-golang/handler.js [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+    err = data.toString('ascii');
+  });
+
+  proc.on('close', (code) => {
+    if (code !== 0) {
+      cb(new Error('Process exited with non-zero status code'));
+    } else if (err) {
+      cb(new Error(err));
+    } else {
+      cb(null, { message: 'Go Serverless v1.0! Your function executed successfully!', eventData });
```
Comment (flomotlik) [inline lib/plugins/create/templates/aws-golang/handler.js]: this should only return event data without the message, otherwise people will always have the message included in their returns

File: lib/plugins/create/templates/aws-golang/handler.js [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+
+  proc.stdout.on('data', (data) => {
+    eventData = JSON.parse(data.toString('ascii'));
+    if (typeof eventData === 'string') {
+      eventData = JSON.parse(eventData);
+    }
+  });
+
+  proc.stderr.on('data', (data) => {
+    err = data.toString('ascii');
```
Comment (flomotlik) [inline lib/plugins/create/templates/aws-golang/handler.js]: This should log from STDERR of the Go tool to STDOUT so Go programs can log

**Comment by license2e** [comment]
@mthenw Added your suggestion/request for external templates: https://github.com/serverless/serverless/issues/1717

File: lib/plugins/create/templates/aws-golang/handler.js (hunk 4 above)
Comment (license2e) [inline lib/plugins/create/templates/aws-golang/handler.js]: @flomotlik  This is meant to copy what is currently in the Nodejs template: https://github.com/serverless/serverless/blob/v1.0/lib/plugins/create/templates/aws-nodejs/handler.js#L5

**Comment by flomotlik** [comment]
Sorry the close was accidental as we removed the V1.0 branch

**Comment by license2e** [comment]
@flomotlik np, I will redo against latest version and with the changes you requested above...

**Comment by flomotlik** [comment]
@license2e I had a conversation with one of our engineers recently and one of the things we talked about is the way Apex for example includes Go. They aren't just putting in the compiled binary, but the binary is actually starting and listens on STDOUT for new data. So the Go binary doesn't run and start for each invocation, but is started once and then runs (so kind of like a Client/Server model). Have you tested the speed of the Go template without this kind of Client/Server model? How fast do the binaries start on Lambda and how much additional time is starting the binary taking?

Because if this is too much I don't think we can really get this out into the Core (even though I like the feature) as it wouldn't be a good experience

**Comment by license2e** [comment]
@flomotlik I have not, unfortunately, my experience with Golang is pretty limited. The template submitted was based on a tutorial I found online. I did take a look at how Apex implemented their Golang setup after I already completed the task, but did not replicate with the template. Perhaps instead of setting up a Golang template in Serverless, we go with @mthenw suggestion with external templates: https://github.com/serverless/serverless/issues/1717 ?
